### DISCUSSION TITLE
How to define entry points

### DISCUSSION THREAD

**Discussion by daknuett** [mesonbuild/meson-python#501] [Q&A] [answered]
The [python packaging specifications](https://packaging.python.org/en/latest/specifications/entry-points/) include _entry points_, most notably `console_scripts` for creating executable console scripts from python packages.  

Old tools, such as `setuptools` have the functionality to define entry points. For meson-python, however, I only found a [stub](https://meson-python.readthedocs.io/en/latest/tutorials/entrypoints.html) in the documentation.

Is there a way to define entry points using meson/meson-python? If yes, are there any pointers on how to use it?

**Comment by dnicolodi** [comment, marked as answer]
Entry points are part of the `pyproject.toml` project metadata specification https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#entry-points `meson-python` implements that specification. There is nothing special you need to do to use entry points with `meson-python` other than following the specification. An example of `pyproject.toml` that defines the the equivalent of `console_scripts` entry points: https://github.com/beancount/beancount/blob/a91c854ea148cda36eb9ea9e8a82f0cc75d40651/pyproject.toml

**Reply by daknuett** [reply]
Thanks. I will make a pull request for updated documentation.

**Reply by rgommers** [reply]
Here's another example, using a couple of `[project.entry-points.*]` entrypoints in addition to `console_scripts`: https://github.com/numpy/numpy/blob/7a23a5406e8b503d60ad811cdae8eb0f4b36fc6d/pyproject.toml#L54-L61

**Reply by daknuett** [reply]
Should the documentation point to the meson build files of other projects? Writing working examples on how to define other entry points appears to be a substantial effort. Or is just mentioning the other entry point groups sufficient?

#502 provides documentation on how to use entry points; so far only `console_scripts`.

**Reply by dnicolodi** [reply]
Linking to the PyPA documentation is more enough IMHO. There is no need for a tutorial. The `setutools` documentation on the same concept also applies https://setuptools.pypa.io/en/latest/userguide/entry_point.html
